### DISCUSSION TITLE
%12X MAC address as variable for MQTT setting and device name

### DISCUSSION THREAD

**Discussion by ghost** [arendst/Tasmota#12515] [Ideas]
At the moment, %12X can be used as variable for MAC address for MQTT_CLIENT_ID and MQTT_TOPIC. Would love it if it could be also used for MQTT_USER and FRIENDLY_NAME.

MQTT_USER that uses its MAC address is mentioned in Securing your Communication section of https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/. This will help automate the setting of each device, instead of having to find its IP, login and enter the MAC address as MQTT_USER manually.

FRIENDLY_NAME is used in Home Assistant as its device name, and subsequently affecting its entity names. When multiple devices are newly connected (and also after reset), they will default to a same fixed pre-compiled name. This affects the entity names, causing it to have _1, _2, _3 etc suffix. Changing them manually is tedious when dealing with many devices. In configuration reset, this will also break the automation, until entity names are corrected manually. When several devices go through this at the same time, it becomes confusing which one is which device.

Hence, when MQTT_USER and FRIENDLY_NAME can be compiled with %12X variable, as like in MQTT_CLIENT_ID and MQTT_TOPIC, this will be a great help. This will help greatly when flashing large number of devices. Also helps when doing device configuration reset.

Hopefully this features request would be looked at favorably.

**Comment by barbudor** [comment]
I think your proposed usage is wrong because it is mixing physical and logical identifications.
The MAC address is a identification of a physical device.
Topic is a logical identification of the function like bedroom/light or bathroom/fan

If you include the MAC address into logical identification, it means that if you have to replace that wall switch or sonoff thing because of a failure, not only you have to configure the replacement device but you also need to reconfigure your home automation.
If you separate physical from logical, you can reconfigure the new device and it will seamlessly integrate in place of the previous one.

Imagine that you include MAC addresses of the server in all hostnames on the Internet.
Like http://google_67A4CE4F

As of today I have the following settings:
ClientID = %12X = MAC address
User = Topic = function identification (like plug-pc1, switch-bedroom, ....)
Mosquitto ACL:
```
pattern read  cmnd/%u/#
pattern read  cmnd/%c_fb/#
pattern write stat/%u/#
pattern write tele/%u/#
pattern read  cmnd/alldevices/#
```
Where %u is the MqttUser set same as Topic in Tasmota and %c in the ClientID set to %12X in Tasmota (fallback topic is always `fb_<ClientID>`

Just my 2c
Doesn't prevent anyone to propose a PR to extend % substitutions in Tasmota

**Reply by ghost** [reply]
> If you include the MAC address into logical identification, it means that if you have to replace that wall switch or sonoff thing because of a failure, not only you have to configure the replacement device but you also need to reconfigure your home automation.

The idea would be:
- flash new device (with default settings, wifi, etc etc, including using MAC address for user, id, etc etc)
- add MAC as user in mosquitto broker
- it will connect to wifi and mqtt automatically
- in HA, it will detect a new device with the MAC as device name, using Tasmota integration (https://www.home-assistant.io/integrations/tasmota/)
- go to entities tab in HA, locate the device entities, change their entity names to meaningful names
- use these meaningful names for automations

When a device fails, just do:
- flash new device (with default settings, wifi, etc etc, including using MAC address for user, id, etc etc)
- add MAC as user in mosquitto broker
- replace the old device with this new device
- it will connect to wifi and mqtt automatically
- in HA, it will detect a new device with the MAC as device name, using Tasmota integration (https://www.home-assistant.io/integrations/tasmota/)
- go to entities tab in HA, locate the device entities, change their entity names to meaningful names (old names)
- everything is back to normal
